core/modules.rs

```diff
@@ -680,13 +683,42 @@ impl Stream for RecursiveModuleLoad {
   }
 }
 
+#[derive(Clone, Copy, Debug, Eq, Hash, PartialEq)]
+pub(crate) enum AssertedModuleType {
+  JavaScriptOrWasm,
+  Json,
+}
+
+// NOTE(bartlomieju): we actually want to have `Into` implementation, instead
+// of `From`, because this operation shouldn't work two-ways, ie. you shouldn't
+// be able to go from `ModuleType` into `AssertedModuleType`. This should
+// probably be removed once WASM module support lands.
+#[allow(clippy::from_over_into)]
+impl Into<AssertedModuleType> for ModuleType {
+  fn into(self) -> AssertedModuleType {
+    match self {
+      ModuleType::JavaScript => AssertedModuleType::JavaScriptOrWasm,
+      ModuleType::Json => AssertedModuleType::Json,
+    }
+  }
+}
```

Comment:
```suggestion
impl From<ModuleType> for AssertedModuleType {
  fn from(module_type: ModuleType) -> AssertedModuleType {
    match module_type {
      ModuleType::JavaScript => AssertedModuleType::JavaScriptOrWasm,
      ModuleType::Json => AssertedModuleType::Json,
    }
  }
}
```

I think you are confusing something here. Regardless of if you use `From` or `Into`, you will only be able to do `ModuleType -> AssertedModuleType`, not the other way around.

The only difference is that with `Into` you can only do `ModuleType::into`, while with `From` you can do `ModuleType::into` and `AssertedModuleType::from`.

Neither of them allow for `AssertedModuleType::into` or `ModuleType::from` (I think this is the one you are concerned about.